### PR TITLE
Add XOAUTH2 authenticator

### PR DESCRIPTION
**Pull request by ssunday**
Thanks to https://github.com/nfo/gmail_xoauth for doing the heavy lifting and changing the license. 💪🏻 

I tested it with a Microsoft 365 account. I would appreciate someone verifying it works with Gmail too to make sure.

Let me know if there's anything else that I can do to get this in 🙏🏻

### CODE REVIEW

**Comment by ssunday** [comment]
Part of https://github.com/ruby/net-imap/issues/12

**Comment by hsbt** [comment]
@ssunday Can you rebase from the current master?

File: lib/net/imap/authenticators/xoauth2.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+  private
+
+  def initialize(user, oauth2_token)
+    @user = user
+    @oauth2_token = oauth2_token
+  end
+
+  def build_oauth2_string(user, oauth2_token)
+    format("user=%s\1auth=Bearer %s\1\1".encode('us-ascii'), user, oauth2_token)
```
Comment (shugo) [inline lib/net/imap/authenticators/xoauth2.rb:16]: What is `.encode('us-ascii')` for?

I use similar code without it, and has no problem with Gmail.

https://github.com/shugo/mournmail/blob/5941cf8c5741b0a7c38470822e5ebc53545caf07/lib/mournmail/utils.rb#L321

Comment (ssunday) [reply lib/net/imap/authenticators/xoauth2.rb:16]: I am not sure the exact reason for it—I haven't tried it without it.

Comment (ssunday) [reply lib/net/imap/authenticators/xoauth2.rb:16]: I can remove it and try it with M365 again

**Comment by ssunday** [comment]
@hsbt rebased

File: lib/net/imap/authenticators/xoauth2.rb (hunk 1 above)
Comment (shugo) [reply lib/net/imap/authenticators/xoauth2.rb:16]: @ssunday Would you remove it?

Comment (ssunday) [reply lib/net/imap/authenticators/xoauth2.rb:16]: Yep.

Comment (ssunday) [reply lib/net/imap/authenticators/xoauth2.rb:16]: I'll try to get to it + testing later today, if not, tomorrow!

**Review by shugo — APPROVED** [review]
It looks fine. Thank you!

**Comment by nevans** [comment]
thanks! I'd started on this a while ago and got completely sidetracked. thanks again!

Additionally, thanks for the reminder that I'd forgotten about it.  No promises, but I'll try to finish up my `ANONYMOUS`, `EXTERNAL`, `OAUTHBEARER`, and `SCRAM-*` PRs this week.

**Comment by Neustradamus** [comment]
@ssunday: Thanks for your PR!

@nevans: Can not wait ^^

**Pull request merged**